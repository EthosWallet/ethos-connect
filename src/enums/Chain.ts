export enum Chain {
  Eth = "ethereum",
  Sui = "sui"
}