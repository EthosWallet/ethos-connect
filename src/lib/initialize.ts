import store from 'store2'

type InitializeOptions = {
  walletAppUrl: string
}

const initialize = ({ walletAppUrl }: InitializeOptions): void => {
  const ethosStore = store.namespace('ethos')
  ethosStore('walletAppUrl', walletAppUrl)
}

export default initialize
