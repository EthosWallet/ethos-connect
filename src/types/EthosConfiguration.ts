export type EthosConfiguration = {
  walletAppUrl: string
  appId: string
  network: string | number
  iframeOrigin?: string
}
