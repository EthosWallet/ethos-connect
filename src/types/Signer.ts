import type { Preapproval } from './Preapproval'
import type { 
  SignableTransaction,
  SuiTransactionResponse
} from '@mysten/sui.js';

export enum SignerType {
  EXTENSION = "extension",
  HOSTED = "hosted"
}

export interface Signer {
  type: SignerType,
  getAccounts: () => Promise<string[]>,
  getAddress: () => Promise<string>,
  signAndExecuteTransaction: (transaction: SignableTransaction) => Promise<SuiTransactionResponse>,
  requestPreapproval: (preApproval: Preapproval) => Promise<boolean>,
  sign: (message: any) => Promise<boolean>,
  disconnect: () => void
}

export interface ExtensionSigner extends Signer {
    type: SignerType.EXTENSION,   
}

export interface HostedSigner extends Signer {
    type: SignerType.HOSTED,
    email?: string,
    logout: () => void
}