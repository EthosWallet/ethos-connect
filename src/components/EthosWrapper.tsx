import React, { useEffect, useState } from 'react'
import { ethers } from 'ethers'
// import {
//   WagmiConfig,
//   createClient,
//   defaultChains,
//   configureChains,
//   useAccount,
//   useProvider,
//   useSigner,
// } from 'wagmi'
// import { Provider } from '../lib/ethersWrapper/Provider'
import getProvider from '../lib/getProvider'

// import { alchemyProvider } from 'wagmi/providers/alchemy'
// import { publicProvider } from 'wagmi/providers/public'

// import { CoinbaseWalletConnector } from 'wagmi/connectors/coinbaseWallet'
// import { InjectedConnector } from 'wagmi/connectors/injected'
// import { MetaMaskConnector } from 'wagmi/connectors/metaMask'
// import { WalletConnectConnector } from 'wagmi/connectors/walletConnect'
// import { Chain } from '../enums/Chain'
import { EthosConfiguration } from 'types/EthosConfiguration'
import initialize from '../lib/initialize'
import useSuiWallet from '../lib/useSuiWallet'

export type ProviderAndSigner = {
  provider: ethers.providers.Web3Provider | any | undefined
  signer: any
}
export interface EthosWrapperProps extends React.HTMLAttributes<HTMLButtonElement> {
  ethosConfiguration: EthosConfiguration
  onProviderSelected: ({ provider, signer }: ProviderAndSigner) => void
}

const EthosWrapper = ({ ethosConfiguration, onProviderSelected, children }: EthosWrapperProps) => {
  console.log('EthosWrapper', ethosConfiguration)
  // const eth = ethosConfiguration.chain === Chain.Eth

  const [providerAndSigner, setProviderAndSigner] = useState<ProviderAndSigner | null>(null)
  const suiProviderAndSigner = useSuiWallet()

  const _onProviderSelected = (providerAndSigner: ProviderAndSigner) => {
    console.log('SET PROVIDER AND SIGNER!!!!!!!', providerAndSigner)
    setProviderAndSigner(providerAndSigner)
    onProviderSelected && onProviderSelected(providerAndSigner)
  }

  const childrenWithProviderAndSigner = React.Children.map(children, (child) => {
    if (React.isValidElement(child)) {
      return React.cloneElement(child, { ...providerAndSigner })
    }
    return child
  })

  // if (eth) {
  //   const ethConfiguration = ethosConfiguration as EthereumConfiguration
  //   const provider = useProvider()
  //   const { address } = useAccount()
  //   const { data: signer } = useSigner()
  //   const {
  //     chains,
  //     provider: chainsProvider,
  //     webSocketProvider,
  //   } = configureChains(defaultChains, [
  //     alchemyProvider({ alchemyId: ethConfiguration.alchemyId }),
  //     publicProvider(),
  //   ])

  //   const client = createClient({
  //     autoConnect: true,
  //     connectors: [
  //       new MetaMaskConnector({ chains }),
  //       // new CoinbaseWalletConnector({
  //       //   chains,
  //       //   options: {
  //       //     appName: 'wagmi',
  //       //   },
  //       // }),
  //       new WalletConnectConnector({
  //         chains,
  //         options: {
  //           qrcode: true,
  //         },
  //       }),
  //       // new InjectedConnector({
  //       //   chains,
  //       //   options: {
  //       //     name: 'Injected',
  //       //     shimDisconnect: true,
  //       //   },
  //       // }),
  //     ],
  //     provider: chainsProvider,
  //     webSocketProvider,
  //   })

  //   useEffect(() => {
  //     if (!address) {
  //       _onProviderSelected({ provider, signer: null })
  //     } else if (signer) {
  //       const fullProvider = new Provider(provider, signer)
  //       _onProviderSelected({ provider: fullProvider, signer })
  //     }
  //   }, [])

  //   return <WagmiConfig client={client}>{childrenWithProviderAndSigner}</WagmiConfig>
  // }

  useEffect(() => {
    console.log(suiProviderAndSigner)
    if (!suiProviderAndSigner) return

    const { provider: suiProvider, signer: suiSigner } = suiProviderAndSigner
    if (suiProvider && !suiSigner) {
      console.log('PROVIDER AND SIGNER1', suiProviderAndSigner)
      initialize(ethosConfiguration)

      const fetchEthosProvider = async () => {
        const ethosProvider = await getProvider()
        _onProviderSelected({
          provider: ethosProvider,
          signer: ethosProvider?.getSigner(),
        })
      }

      fetchEthosProvider()
    } else if (suiSigner) {
      console.log('PROVIDER AND SIGNER2', suiProviderAndSigner)
      _onProviderSelected(suiProviderAndSigner)
    }
  }, [suiProviderAndSigner])

  return <>{childrenWithProviderAndSigner}</>
}

export default EthosWrapper
