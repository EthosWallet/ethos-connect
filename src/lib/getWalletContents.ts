import { Connection, JsonRpcProvider } from "@mysten/sui.js";
import { SuiNFT, WalletContents } from "../types/WalletContents";
import { newBN, sumBN } from './bigNumber';
import getBagNFT, { isBagNFT } from "./getBagNFT";
// import fetchSui from "./fetchSui";
import { ConvenenienceSuiObject } from '../types/ConvenienceSuiObject';
import { DEFAULT_NETWORK } from './constants';
import getDisplay from "./getDisplay";

export const ipfsConversion = (src?: string): string => {
    if (!src) return "";
    if (src.indexOf('ipfs') === 0) {  
        src = `https://ipfs.io/ipfs/${src.substring(5)}`;
    }
    return src;
}

export type GetWalletContentsArgs = {
    address: string,
    network: string,
    existingContents?: WalletContents
}

const empty: WalletContents = {
    suiBalance: newBN(0),
    nfts: [],
    tokens: {},
    objects: []  
}

const getWalletContents = async ({ address, network, existingContents }: GetWalletContentsArgs): Promise<WalletContents | null> => {
    try {
        const connection = new Connection({ fullnode: network || DEFAULT_NETWORK })
        const provider = new JsonRpcProvider(connection);

        if (!address) {
            return empty
        }
        
        const objectInfos = await provider.getOwnedObjects({
            owner: address,
            options: {
                showType: true,
                showOwner: true,
                showContent: true,
                showDisplay: true,
            }
        });

        if (objectInfos.data.length === 0) {
            if (existingContents === empty) {
                return null;
            }
            return empty;
        }

        const currentObjects = [];
        let newObjectInfos = [];
        if (existingContents?.objects && existingContents.objects.length > 0) {
            for (const objectInfo of objectInfos.data) {
                if (!objectInfo.data || objectInfo.error) continue;
                const existingObject = existingContents?.objects.find(
                    (existingObject) => {
                        if (
                            typeof objectInfo.data === "object" &&
                            typeof existingObject.data === "object"
                        ) {
                            return (
                                existingObject.data.objectId === objectInfo.data.objectId &&
                                existingObject.data.version === objectInfo.data.version
                            )    
                        } else {
                            return false;
                        }
                    }
                );
                
                if (existingObject) {
                    currentObjects.push(existingObject);
                } else {
                    newObjectInfos.push(objectInfo);
                }
            }
        } else {
            newObjectInfos = objectInfos.data;
        }

        if (newObjectInfos.length === 0) return null;

        const newObjects = newObjectInfos;
        // const newObjects = newObjectInfos.map((o) => {
        //     if (typeof o.data === "object") {
        //         return o.data.objectId
        //     } else {
        //         return ""
        //     }
        // }).filter((objectId) => objectId.length > 0);
        
        // const newObjects = await provider.multiGetObjects({
        //     ids: newObjectIds, 
        //     options: {
        //         showContent: true,
        //         showType: true,
        //         showOwner: true,
        //         showDisplay: true
        //     }
        // });
        const objects = currentObjects.concat(newObjects);

        let suiBalance = newBN(0);
        const nfts: SuiNFT[] = [];
        const tokens: {[key: string]: any}= {};
        const convenenienceObjects: ConvenenienceSuiObject[] = [];
        for (const object of objects) {
            const { data } = object
            if (!data) continue;

            const { display, content: { fields } } = data;
            const safeDisplay = getDisplay(display);
            try {
                const typeStringComponents = (data.type || "").split('<');
                const subtype = (typeStringComponents[1] || "").replace(/>/, '')
                const typeComponents = typeStringComponents[0].split('::');
                const type = typeComponents[typeComponents.length - 1];

                const { name, description, ...extraFields } = fields ?? {}
                
                convenenienceObjects.push({
                    ...object,
                    type: data?.type,
                    version: data?.version,
                    objectId: data?.objectId,
                    name,
                    description,
                    display: safeDisplay,
                    extraFields
                })

                if (type === 'Coin') {
                    if (subtype === '0x2::sui::SUI') {
                        suiBalance = sumBN(suiBalance, fields.balance);
                    }
                    
                    tokens[subtype] ||= {
                        balance: 0,
                        coins: []
                    }
                    
                    tokens[subtype].balance = sumBN(tokens[subtype].balance, fields.balance);
                    tokens[subtype].coins.push({
                        objectId: data?.objectId,
                        type: data?.type,
                        balance: newBN(fields.balance),
                        digest: data?.digest,
                        version: data?.version,
                        display: safeDisplay
                    })
                } else if (isBagNFT(object.data)) {
                    const bagNFT = await getBagNFT(provider, object.data);
                    
                    if ("name" in bagNFT) {
                        nfts.push({
                            type: data?.type,
                            package: typeComponents[0],
                            chain: 'Sui',
                            address: data?.objectId,
                            objectId: data?.objectId,
                            name: safeDisplay?.name ?? bagNFT.name,
                            description: safeDisplay?.name ?? bagNFT.description,
                            imageUri: ipfsConversion(safeDisplay?.image_url ?? bagNFT.url),
                            link: safeDisplay?.link,
                            creator: safeDisplay?.creator,
                            projectUrl: safeDisplay?.project_url,
                            display: safeDisplay,
                            module: typeComponents[1],
                            links: {
                                'Explorer': `https://explorer.sui.io/objects/${object?.objectId}`
                            }
                        });       
                    }
                } else {
                    const { url, image_url, image, ...remaining } = extraFields || {}
                    const safeUrl = ipfsConversion(safeDisplay?.image_url || url || image_url || image);
                    if (safeUrl) {
                        nfts.push({
                            type: data?.type,
                            package: typeComponents[0],
                            chain: 'Sui',
                            address: data?.objectId,
                            objectId: data?.objectId,
                            name: safeDisplay?.name ?? name,
                            description: safeDisplay?.description ?? description,
                            imageUri: safeUrl,
                            link: safeDisplay?.link,
                            creator: safeDisplay?.creator,
                            projectUrl: safeDisplay?.project_url,
                            display: safeDisplay,
                            extraFields: remaining,
                            module: typeComponents[1],
                            links: {
                                'Explorer': `https://explorer.sui.io/objects/${object?.objectId}`
                            }
                        });    
                    }
                }
            } catch (error: unknown) {
                console.log("Error retrieving object", object, error);
            }
        } 

        return { suiBalance, tokens, nfts, objects: convenenienceObjects }
    } catch (error: unknown) {
        console.log("Error retrieving wallet contents", error);
        return empty;
    }
}

export default getWalletContents;