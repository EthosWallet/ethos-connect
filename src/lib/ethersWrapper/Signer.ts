import { TransactionRequest, Provider } from '@ethersproject/abstract-provider';
import { Bytes, Signer as EthersSigner } from 'ethers';
import { Deferrable } from 'ethers/lib/utils';

export class Signer extends EthersSigner {
    getAddress(): Promise<string> {
        return Promise.resolve('THIS IS AN ADDRESS');
    }
    signMessage(message: string | Bytes): Promise<string> {
        throw new Error('Method not implemented.');
    }
    signTransaction(transaction: Deferrable<TransactionRequest>): Promise<string> {
        throw new Error('Method not implemented.');
    }
    connect(provider: Provider): Signer {
        throw new Error('Method not implemented.');
    }

}