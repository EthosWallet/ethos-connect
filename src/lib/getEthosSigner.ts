import { SignableTransaction, SuiTransactionResponse } from '@mysten/sui.js'
import store from 'store2'
import { Chain } from '../enums/Chain'
import { Signer, SignerType } from '../types/Signer'
import activeUser from './activeUser'
import hostedInteraction, { HostedInteractionResponse } from './hostedInteraction'


const getEthosSigner = async (): Promise<Signer | null> => {

    const user: any = await activeUser()
    
    const getAccounts = async () => {
        if (!user) return [];
        return user.accounts.filter((account: any) => account.chain === Chain.Sui)
    }

    const getAddress = async () => {
        const accounts = await getAccounts();
        return accounts[0]?.address;
    }

    const signAndExecuteTransaction = (details: SignableTransaction): Promise<SuiTransactionResponse> => {
        return new Promise((resolve, reject) => {
            const transactionEventListener = ({ approved, data }: HostedInteractionResponse) => {
                if (approved) {
                    resolve(data);
                } else {
                    reject({ error: "User rejected transaction."})
                }
            }
        
            hostedInteraction({
                action: 'transact',
                data: { details },
                onResponse: transactionEventListener
            })
        });
    }

    const requestPreapproval = () => {
        return Promise.resolve(true);
    }

    const sign = () => {
        return Promise.resolve(true);
    }

    const disconnect = (wallet = false) => {
        return new Promise((resolve) => {
            const transactionEventListener = () => {
                resolve(true);
            }
        
            hostedInteraction({
                action: 'logout',
                data: { wallet },
                onResponse: transactionEventListener
            })

            store.namespace('auth')('access_token', null)
        });
    }

    return user ? {
        type: SignerType.ETHOS_HOSTED,
        email: user.email,
        getAccounts,
        getAddress,
        signAndExecuteTransaction,
        requestPreapproval,
        sign,
        disconnect
    } : null
   
}

export default getEthosSigner;