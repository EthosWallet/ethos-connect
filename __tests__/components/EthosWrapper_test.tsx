import React from 'react'
import { create, act } from 'react-test-renderer'

import EthosWrapper from '../../src/components/EthosWrapper'
import getWalletContents from '../../src/lib/getWalletContents'
import { Chain } from '../../src/enums/Chain'
import { EthosConfiguration } from '../../src/types/EthosConfiguration'
import initialize from '../../src/lib/initialize';
import getEthosSigner from '../../src/lib/getEthosSigner'
import { SignerType } from '../../src/types/Signer'

const spyMethods = { getWalletContents, initialize, getEthosSigner }

describe('EthosWrapper', () => {
  const signer = {
    getAddress: () => "ADDRESS"
  }

  let receivedProvider
  let receivedSigner
  let onWalletConnected

  beforeEach(() => {
    jest.spyOn(spyMethods, 'getEthosSigner').mockImplementation(() => {
      return Promise.resolve({
        signer,
        getAddress: () => Promise.resolve("ADDRESS"),
        getAccounts: () => Promise.resolve([]),
        type: SignerType.EXTENSION,
        signAndExecuteTransaction: (_transaction) => Promise.resolve({} as any),
        requestPreapproval: (_preApproval) => Promise.resolve(true),
        sign: (_message) => Promise.resolve(true),
        disconnect: () => {}
      })
    })

    onWalletConnected = jest.fn(({ provider: p, signer: s }) => {
      receivedProvider = p
      receivedSigner = s
    });

    jest.spyOn(getWalletContents as any, 'default').mockReturnValue({
      balance: 0,
      coins: [],
      nfts: []
    })
  })

  afterEach(() => {
    receivedProvider = null
    receivedSigner = null
  })

  it('renders nothing but the children provided', async () => {
    let ethosWrapper
    await act(async () => {
      ethosWrapper = create(
        <EthosWrapper ethosConfiguration={{}} onWalletConnected={onWalletConnected}>
          test
        </EthosWrapper>
      )
    })

    expect(ethosWrapper.toJSON()).toMatchSnapshot()
  })

  it('renders calls the onWalletConnected callback', async () => {
    let ethosWrapper
    await act(async () => {
      ethosWrapper = create(
        <EthosWrapper ethosConfiguration={{}} onWalletConnected={onWalletConnected}>
          test
        </EthosWrapper>
      )
    })

    expect(onWalletConnected.mock.calls.length).toBe(1)
    expect(receivedProvider.getSigner()).toBe(signer)
    expect(receivedSigner).toBe(signer)
  })

  it('should initialize default configuration if no optional values are given', async () => {
    let ethosWrapper: any;
    const initialEthosConfiguration: EthosConfiguration = { appId: 'test-id' }
    const expectedEthosConfiguration: EthosConfiguration = {
      appId: 'test-id',
      walletAppUrl: 'https://ethoswallet.xyz',
      chain: Chain.Sui,
      network: 'sui'
    }

    const initializeSpy = jest.spyOn(spyMethods, 'initialize')

    await act(async () => {
      ethosWrapper = create(
        <EthosWrapper ethosConfiguration={initialEthosConfiguration} onWalletConnected={onWalletConnected}>
          test
        </EthosWrapper>
      )
    })

    expect(initializeSpy).toBeCalledWith(expectedEthosConfiguration)
  });
})
