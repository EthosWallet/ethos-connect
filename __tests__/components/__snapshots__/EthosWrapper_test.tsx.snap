// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthosWrapper renders nothing but the children provided 1`] = `
Array [
  "test",
  <link
    href="https://rsms.me/inter/inter.css"
    rel="stylesheet"
  />,
  <div
    style={
      Object {
        "color": "black",
        "fontFamily": "'Inter', sans-serif",
        "fontSize": "16px",
        "lineHeight": "1.5",
      }
    }
  >
    <div
      role="dialog"
      style={
        Object {
          "position": "relative",
          "visibility": "hidden",
          "zIndex": "100",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "rgb(107 114 128)",
            "bottom": "0px",
            "left": "0px",
            "opacity": "0",
            "position": "fixed",
            "right": "0px",
            "top": "0px",
            "transition": "all 300ms ease-in-out",
          }
        }
      />
      <div
        style={
          Object {
            "bottom": "0px",
            "left": "0px",
            "opacity": "0",
            "overflowY": "auto",
            "position": "fixed",
            "right": "0px",
            "scale": ".95",
            "top": "0px",
            "transition": "all 300ms ease-in-out",
            "zIndex": "99",
          }
        }
      >
        <div
          id="ethos-close-on-click"
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
              "minHeight": "100%",
              "padding": "0",
              "textAlign": "center",
            }
          }
        >
          <div
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderRadius": "0.5rem",
                "boxShadow": "0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)",
                "maxWidth": "400px",
                "overflow": "hidden",
                "paddingBottom": "32px",
                "position": "relative",
                "transitionProperty": "all",
                "width": "100%",
              }
            }
          >
            <div
              style={
                Object {
                  "height": "36px",
                }
              }
            >
              <span
                onClick={[Function]}
                style={
                  Object {
                    "backgroundColor": "#F9FAFB",
                    "borderRadius": "100%",
                    "color": "#A0AEBA",
                    "cursor": "pointer",
                    "float": "right",
                    "height": "24px",
                    "marginRight": "12px",
                    "marginTop": "12px",
                    "textAlign": "center",
                    "width": "24px",
                  }
                }
              >
                âœ•
              </span>
            </div>
            <div
              style={
                Object {
                  "padding": "0 30px",
                }
              }
            >
              <div>
                <div
                  style={
                    Object {
                      "borderBottom": "1px solid rgb(241 245 249)",
                      "display": "flex",
                      "flexDirection": "column",
                      "gap": "12px",
                      "justifyContent": "center",
                      "padding": "12px",
                      "width": "100%",
                    }
                  }
                >
                  <div
                    style={
                      Object {
                        "display": "flex",
                        "gap": "-6px",
                        "justifyContent": "center",
                      }
                    }
                  />
                  <div
                    style={
                      Object {
                        "fontSize": "24px",
                        "fontWeight": "600",
                        "lineHeight": "32px",
                        "margin": "0",
                      }
                    }
                  />
                  <div
                    style={
                      Object {
                        "fontSize": "16px",
                        "fontWeight": "400",
                        "lineHeight": "24px",
                        "margin": "0",
                      }
                    }
                  >
                    Log in to 
                  </div>
                </div>
                <div
                  role="email-sign-in"
                >
                  <form
                    onSubmit={[Function]}
                  >
                    <input
                      onChange={[Function]}
                      placeholder="Email address"
                      style={
                        Object {
                          "border": "1px solid rgb(203 213 225)",
                          "borderRadius": "0.5rem",
                          "boxSizing": "border-box",
                          "marginTop": "12px",
                          "padding": "12px",
                          "width": "100%",
                        }
                      }
                      type="email"
                      value=""
                    />
                    <button
                      style={
                        Object {
                          "backgroundColor": "#761AC7",
                          "border": "1px solid rgb(203 213 225)",
                          "borderRadius": "0.5rem",
                          "color": "#FFFFFF",
                          "cursor": "pointer",
                          "fontWeight": "600",
                          "marginTop": ".75rem",
                          "padding": "12px 20px 12px 20px",
                          "textDecoration": "none",
                          "width": "100%",
                        }
                      }
                      type="submit"
                    >
                      Sign In
                    </button>
                  </form>
                  <div
                    style={
                      Object {
                        "marginLeft": "-12px",
                      }
                    }
                  >
                    <div />
                  </div>
                </div>
                <div
                  style={
                    Object {
                      "borderTop": "1px solid #0000001F",
                      "color": "#74777C",
                      "display": "flex",
                      "justifyContent": "center",
                      "margin": "44px 0 30px 0",
                    }
                  }
                >
                  <div
                    style={
                      Object {
                        "backgroundColor": "white",
                        "marginTop": "-14px",
                        "padding": "0 6px",
                      }
                    }
                  >
                    or
                  </div>
                </div>
                <button
                  onClick={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "background": "#F2F2F2",
                      "borderRadius": "12px",
                      "boxSizing": "border-box",
                      "color": "black",
                      "display": "flex",
                      "flex": "none",
                      "flexDirection": "row",
                      "flexGrow": "0",
                      "gap": "10px",
                      "height": "48px",
                      "justifyContent": "space-between",
                      "order": "0",
                      "padding": "24px",
                      "textDecoration": "none",
                      "width": "100%",
                    }
                  }
                >
                  <div>
                    Install A Wallet
                  </div>
                  <svg
                    fill="none"
                    height={42}
                    viewBox="0 0 56 56"
                    width={42}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      filter="url(#filter0_d_514_2149)"
                    >
                      <rect
                        fill="url(#paint0_angular_514_2149)"
                        height="32"
                        rx="8"
                        shapeRendering="crispEdges"
                        width="32"
                        x="12"
                        y="8"
                      />
                      <path
                        d="M19 18.25C19.6485 17.7621 20.4384 17.4988 21.25 17.5H34.75C35.594 17.5 36.373 17.779 37 18.25C37 17.9545 36.9418 17.6619 36.8287 17.389C36.7157 17.116 36.5499 16.8679 36.341 16.659C36.1321 16.4501 35.884 16.2843 35.611 16.1713C35.3381 16.0582 35.0455 16 34.75 16H21.25C20.6533 16 20.081 16.2371 19.659 16.659C19.2371 17.081 19 17.6533 19 18.25V18.25ZM19 21.25C19.6485 20.7621 20.4384 20.4988 21.25 20.5H34.75C35.594 20.5 36.373 20.779 37 21.25C37 20.9545 36.9418 20.6619 36.8287 20.389C36.7157 20.116 36.5499 19.8679 36.341 19.659C36.1321 19.4501 35.884 19.2843 35.611 19.1713C35.3381 19.0582 35.0455 19 34.75 19H21.25C20.6533 19 20.081 19.2371 19.659 19.659C19.2371 20.081 19 20.6533 19 21.25V21.25ZM25 22C25.2652 22 25.5196 22.1054 25.7071 22.2929C25.8946 22.4804 26 22.7348 26 23C26 23.5304 26.2107 24.0391 26.5858 24.4142C26.9609 24.7893 27.4696 25 28 25C28.5304 25 29.0391 24.7893 29.4142 24.4142C29.7893 24.0391 30 23.5304 30 23C30 22.7348 30.1054 22.4804 30.2929 22.2929C30.4804 22.1054 30.7348 22 31 22H34.75C35.0455 22 35.3381 22.0582 35.611 22.1713C35.884 22.2843 36.1321 22.4501 36.341 22.659C36.5499 22.8679 36.7157 23.116 36.8287 23.389C36.9418 23.6619 37 23.9545 37 24.25V29.75C37 30.0455 36.9418 30.3381 36.8287 30.611C36.7157 30.884 36.5499 31.1321 36.341 31.341C36.1321 31.5499 35.884 31.7157 35.611 31.8287C35.3381 31.9418 35.0455 32 34.75 32H21.25C20.6533 32 20.081 31.7629 19.659 31.341C19.2371 30.919 19 30.3467 19 29.75V24.25C19 23.6533 19.2371 23.081 19.659 22.659C20.081 22.2371 20.6533 22 21.25 22H25Z"
                        fill="white"
                      />
                    </g>
                    <defs>
                      <filter
                        colorInterpolationFilters="sRGB"
                        filterUnits="userSpaceOnUse"
                        height="56"
                        id="filter0_d_514_2149"
                        width="56"
                        x="0"
                        y="0"
                      >
                        <feFlood
                          floodOpacity="0"
                          result="BackgroundImageFix"
                        />
                        <feColorMatrix
                          in="SourceAlpha"
                          result="hardAlpha"
                          type="matrix"
                          values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0"
                        />
                        <feOffset
                          dy="4"
                        />
                        <feGaussianBlur
                          stdDeviation="6"
                        />
                        <feComposite
                          in2="hardAlpha"
                          operator="out"
                        />
                        <feColorMatrix
                          type="matrix"
                          values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.16 0"
                        />
                        <feBlend
                          in2="BackgroundImageFix"
                          mode="normal"
                          result="effect1_dropShadow_514_2149"
                        />
                        <feBlend
                          in="SourceGraphic"
                          in2="effect1_dropShadow_514_2149"
                          mode="normal"
                          result="shape"
                        />
                      </filter>
                      <radialGradient
                        cx="0"
                        cy="0"
                        gradientTransform="translate(50.5 36) rotate(52.1843) scale(42.4058)"
                        gradientUnits="userSpaceOnUse"
                        id="paint0_angular_514_2149"
                        r="1"
                      >
                        <stop
                          offset="0.32986"
                          stopColor="#E57948"
                        />
                        <stop
                          offset="0.396972"
                          stopColor="#4AC2A7"
                        />
                        <stop
                          offset="0.441723"
                          stopColor="#65B1DF"
                        />
                        <stop
                          offset="0.492932"
                          stopColor="#BCA1E8"
                        />
                        <stop
                          offset="0.570879"
                          stopColor="#EED32C"
                        />
                      </radialGradient>
                    </defs>
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
